Source/WebGPU/WebGPU/BindGroup.mm

```diff
@@ -179,31 +79,35 @@ static auto sizeOfEntries(const WGPUBindGroupDescriptor& descriptor, BindGroupLa
+        if (bufferIsPresent + samplerIsPresent + textureViewIsPresent != 1)
             return BindGroup::createInvalid(*this);
 
-        if (bufferIsPresent) {
-            id<MTLBuffer> buffer = WebGPU::fromAPI(entry.buffer).buffer();
-            [vertexArgumentEncoder setBuffer:buffer offset:static_cast<NSUInteger>(entry.offset) atIndex:entry.binding];
-            [fragmentArgumentEncoder setBuffer:buffer offset:static_cast<NSUInteger>(entry.offset) atIndex:entry.binding];
-            [computeArgumentEncoder setBuffer:buffer offset:static_cast<NSUInteger>(entry.offset) atIndex:entry.binding];
-        } else if (samplerIsPresent) {
-            id<MTLSamplerState> sampler = WebGPU::fromAPI(entry.sampler).samplerState();
-            [vertexArgumentEncoder setSamplerState:sampler atIndex:entry.binding];
-            [fragmentArgumentEncoder setSamplerState:sampler atIndex:entry.binding];
-            [computeArgumentEncoder setSamplerState:sampler atIndex:entry.binding];
-        } else if (textureViewIsPresent) {
-            id<MTLTexture> texture = WebGPU::fromAPI(entry.textureView).texture();
-            [vertexArgumentEncoder setTexture:texture atIndex:entry.binding];
-            [fragmentArgumentEncoder setTexture:texture atIndex:entry.binding];
-            [computeArgumentEncoder setTexture:texture atIndex:entry.binding];
-        } else {
-            ASSERT_NOT_REACHED();
-            return BindGroup::createInvalid(*this);
+        auto stages = bindGroupLayout.stagesForBinding(entry.binding);
+        for (size_t currentStage = 0; currentStage < shaderStageLength; ++currentStage) {
+            WGPUShaderStage renderStage = shaderStage[currentStage];
+            if (!(stages & renderStage))
+                continue;
+
+            auto& index = bindingOffsetForStage[currentStage];
+            if (bufferIsPresent) {
+                id<MTLBuffer> buffer = WebGPU::fromAPI(entry.buffer).buffer();
+                if (entry.offset > buffer.length)
+                    return BindGroup::createInvalid(*this);
+
+                [argumentEncoder[currentStage] setBuffer:buffer offset:entry.offset atIndex:index++];
+            } else if (samplerIsPresent) {
+                id<MTLSamplerState> sampler = WebGPU::fromAPI(entry.sampler).samplerState();
+                [argumentEncoder[currentStage] setSamplerState:sampler atIndex:index++];
+            } else if (textureViewIsPresent) {
+                id<MTLTexture> texture = WebGPU::fromAPI(entry.textureView).texture();
+                [argumentEncoder[currentStage] setTexture:texture atIndex:index++];
+                resources.append({ texture, MTLResourceUsageRead, renderStage });
+            }
         }
     }
 
-    return BindGroup::create(vertexArgumentBuffer, fragmentArgumentBuffer, computeArgumentBuffer, WTFMove(resources), *this);
+    constexpr int vertex = 0, fragment = 1, compute = 2;
```

Comment:
Right, the enum would save you from having to do this

Reply:
I will add the enum to decouple from WGPU* types but indexing with an enum, in addition to a cast, requires the assumption the enum values are 0,1,2 and not a bitfield or some other values too? I could static_assert to guarantee that, but seems like until this becomes more complicated `constexpr int vertex = 0, fragment = 1, compute = 2;` is easier / more readable and addresses @djg's concern from a previous PR about indexing with [0], [1], [2]